FFmpegInterop/NALUSampleProvider.cpp

```diff
@@ -0,0 +1,255 @@
+
+		// Transform the sample into the format expected by the decoder
+		auto [buf, naluLengths] = TransformSample(move(packet), isKeyFrame);
+
+		// Set sample properties
+		map<GUID, IInspectable> properties;
+
+		if (isKeyFrame)
+		{
+			properties[MFSampleExtension_CleanPoint] = PropertyValue::CreateUInt32(true);
+		}
+
+		const uint8_t* naluLengthsBuf{ reinterpret_cast<const uint8_t*>(naluLengths.data()) };
+		const size_t naluLengthsBufSize{ sizeof(decltype(naluLengths)::value_type) * naluLengths.size() };
+		properties[MF_NALU_LENGTH_INFORMATION] = PropertyValue::CreateUInt8Array({ naluLengthsBuf, naluLengthsBuf + naluLengthsBufSize });
+
+		return { move(buf), pts, dur, move(properties) };
+	}
+
+	tuple<IBuffer, vector<uint32_t>> NALUSampleProvider::TransformSample(_Inout_ AVPacket_ptr packet, _In_ bool isKeyFrame)
+	{
+		bool copySpsPpsData{ isKeyFrame && !m_spsPpsData.empty() };
+		const bool writeToBuf{ copySpsPpsData || (!m_isBitstreamAnnexB && m_naluLengthSize != sizeof(NALU_START_CODE)) };
+
+		vector<uint8_t> buf;
+		if (writeToBuf)
+		{
+			size_t sampleSize{ static_cast<uint32_t>(packet->size) };
+
+			if (isKeyFrame)
+			{
+				// Reserve space for SPS/PPS data
+				sampleSize += m_spsPpsData.size();
+			}
+
+			if (!m_isBitstreamAnnexB && m_naluLengthSize > sizeof(NALU_START_CODE))
+			{
+				// Reserve additional space since NALU start code is longer than NALU length size
+				sampleSize += MAX_NALU_NUM_SUPPORTED * (sizeof(NALU_START_CODE) - m_naluLengthSize);
+			}
+
+			buf.reserve(sampleSize);
+		}
+
+		vector<uint32_t> naluLengths;
+		naluLengths.reserve(MAX_NALU_NUM_SUPPORTED);
+
+		const uint8_t naluPrefixLength{ m_isBitstreamAnnexB ? sizeof(NALU_START_CODE) : m_naluLengthSize };
+		for (uint32_t i{ 0 }; i + naluPrefixLength <= static_cast<uint32_t>(packet->size);)
+		{
+			// Get the length of the next NALU
+			uint32_t naluLength{ 0 };
+			if (m_isBitstreamAnnexB)
+			{
+				naluLength = GetAnnexBNaluLength(packet->data + i, packet->size - i);
+			}
+			else
+			{
+				naluLength = GetAVCNaluLength(packet->data + i, packet->size - i, m_naluLengthSize);
+			}
+
+			if (copySpsPpsData)
+			{
+				// Check if this NALU is an AUD (access unit delimiter). The AUD must be the first NALU in the sample.
+				if (naluLength < naluPrefixLength || packet->data[naluPrefixLength] != NALU_TYPE_AUD)
+				{
+					// Copy the SPS/PPS data
+					buf.insert(buf.end(), m_spsPpsData.data(), m_spsPpsData.data() + m_spsPpsData.size());
+
+					// Save the SPS/PPS NALU lengths
+					naluLengths.insert(naluLengths.end(), m_spsPpsNaluLengths.begin(), m_spsPpsNaluLengths.end());
+
+					copySpsPpsData = false;
+				}
+			}
+
+			if (writeToBuf)
+			{
+				// Write the NALU start code and data
+				const uint8_t* naluData{ packet->data + i + naluPrefixLength };
+				buf.insert(buf.end(), begin(NALU_START_CODE), end(NALU_START_CODE));
+				buf.insert(buf.end(), naluData, naluData + naluLength);
```

Comment:
>buf.insert(buf.end(), naluData, naluData + naluLength); [](start = 4, length = 55)

Will this throw if naluLength is 0? If so, we should probably have an explicit check here to skip the insert call since we wouldn't want the presence of empty NALUs to cause fatal failures.

Reply:
No it's safe. It's just a no-op when naluLength is 0.